### ISSUE TITLE
Hold key doesn't work as expected in tty

### ISSUE THREAD

**Issue by timsofteng** [kmonad/kmonad#543]
Hello. When i hold `h` key in my vim or file manager I expect scrolling to bottom.
With enabled kmonad it works as expected in `sway` wm but it doesnt in tty. It does only one jump and stop.

How to fix it?

**Comment by MuhammedZakir** [comment]
Key-repeat doesn't work in TTY. KMonad-native key-repeat is only implemented for Windows. In Linux, X/Wayland is doing the key-repeating. That's why you don't get it in TTY.

I am still confused on why there is key-repeat for physical keyboards, but not for virtual ones? Has anyone used any other low-level key-remapping tool? Does it has key-repeat in TTY?

**Comment by david-janssen** [comment]
> Key-repeat doesn't work in TTY.

Ahhhh, I should fix that. I'm putting it on the list!

> I am still confused on why there is key-repeat for physical keyboards, but not for virtual ones? 

I am not sure myself. I just tried running `evtest` on my simulated `uinput` device, but that doesn't want to cooperate. I know that when I implemented the linux event injection, key-repeat was working automatically when I was testing stuff in X11. I never thought to test in the TTY, so for the sake of simplicity we only inject `KeyPress` and `KeyRelease` events.

Just googled some stuff, and I'm guessing that the reason key-repeat was working was because X11 has its own key-repeat mechanism that was taking care of this. The fix would be to add key-repeat into the uinput sink. This is doable, but ideally I'd generalize the key-repeat from Windows and then use that 1 module in both Windows and Linux. Since TTY stuff is not used too much, I'm not going to give this a lot of priority. There are some other things I want to work on first, but I will try to get to this.

**Comment by timsofteng** [comment]
@david-janssen cool thanks.
Please notice that it works in Linux tty without kmonad.

**Comment by MuhammedZakir** [comment]
> Please notice that it works in Linux tty without kmonad.

Yes. "Normal" keyboard works fine. "Virtual" keyboards don't. One thing to make sure is whether this has to do with KMonad itself, or `uinput` in general. A list of low-level remapping tools I know of [1]. We can test these to see whether it supports key-repeat in TTY without any app-native key-repeat mechanism.

[1] https://github.com/kmonad/kmonad/discussions/388#discussioncomment-1658407

---

@david-janssen: There is some key-repeat related stuff in both `linux/input.h` and `linux/input-event-codes.h` include files. Search for `repeat` in former and `Autorepeat` in latter file. Are those relevant?

**Comment by david-janssen** [comment]
> Please notice that it works in Linux tty without kmonad.

Yes, I am aware :-). This lines up with my working hypothesis:
- Linux kernel deals in Push/Repeat/Release events
- X11 has its own repeat mechanism
- When I tested KMonad, X11 repeat mechanism kicked in, and I didn't realize it was X11 caused
- Therefore I never modelled kernel-level key-repeat

I need to go into KMonad and add a representation of kernel-level key-repeat to the Linux output sink. That should do it. I am in the middle of dealing with some keycode stuff, but I'll try and get to this soon after, or when I feel very uninspired on the keycode stuff and I just want to deal with a simple but tedious task.
